Docs/conductor documentation (#302)
astriaorg/astria: specs/conductor.md

# Astria Conductor

> The Astria Conductor connects the shared Sequencer and data availability layers to the execution layer. There is one instance of the Conductor per rollup.

Reviewer: ```suggestion
The Astria Conductor connects the shared sequencer and data availability layers to the execution layer. There is one instance of the Conductor per rollup node.
```

Reviewer: not even sure if "rollup node" is the right term here, it's more like "rollup execution environment" while "rollup node" in our stack would refer to the combo of "rollup execution environment + conductor"

Blocks can be received via either:

- The gossip network from the shared Sequencer
- The data availability layer, requested on a predefined interval

In the first case, the transactions in the block are filtered and pushed to the execution layer, executed, and added to the blockchain. Brand new blocks are marked as head and all previous blocks are marked as a soft commitment. In general, the marking of blocks as safe will happen one block at a time as new blocks arrive. The block is not finalized until it's received from the data availability layer.

In the second case, batches of blocks are received from the DA layer and filtered for the rollup. These blocks are then used to set their corresponding blocks' commit status on the rollup as finalized.

The exact terminology that a rollup uses for its fork choice rules is up to its implementation. For example, Geth uses `head`, `safe`, and `final`. The Conductor uses `head`, `soft`, and `firm`. The fork choice options are mapped with Geth in the following way:

- `head` -> `head`
- `soft` -> `safe`
- `firm` -> `final`

> The Conductor has no knowledge that Geth uses the above options, it simply sends a `CommitmentState` message to the rollup, and the rollup does the rest.

Reviewer: ```suggestion
To update the commitment level of a block on the rollup, it simply sends a `CommitmentState` message to the rollup node.
```

    - The gossip network uses the [astria-gossipnet](https://github.com/astriaorg/astria/tree/main/crates/astria-gossipnet)
- Runs an event loop that handles receiving `DriverCommand`s from other actors as well as messages from the gossip network
- Validates and passes Sequencer blocks received from the gossip network to the `Executor`

The Driver receives either `Events` ([link](https://github.com/astriaorg/astria/blob/6e71a76fa52c522ffdcabcd9d659e4de765d9d61/crates/astria-gossipnet/src/network_stream.rs#L39)) from the network, or `DriverCommand`s ([link](https://github.com/astriaorg/astria/blob/6e71a76fa52c522ffdcabcd9d659e4de765d9d61/crates/astria-conductor/src/driver.rs#L54)) from
the Conductor's internal event loop on a timer. The variants for `Event` and
`DriverCommand` that are relevant to the processing of blocks within the
Conductor are:

- `Event::GossipsubMessage(Message)` ([link](https://github.com/astriaorg/astria/blob/6e71a76fa52c522ffdcabcd9d659e4de765d9d61/crates/astria-gossipnet/src/network_stream.rs#L50))
    - The `Message` value within the `GossipsubMessage` contains the Sequencer
      block data from the Astria Sequencer value for an `Event` is received the

>       `Message` is then parsed to a `SequencerBlockData`

Reviewer 2: I think this should say:
The `Message` value within the `GossipsubMessage` contains the Sequencer block data from the Astria Sequencer. The `Message` is then parsed to a `SequencerBlockData`

      ([link](https://github.com/astriaorg/astria/blob/6e71a76fa52c522ffdcabcd9d659e4de765d9d61/crates/astria-sequencer-types/src/sequencer_block_data.rs#L39)).

>       Once transformed, the block data is validated to make sure that the
>       validator set correctly agreed on the block, the proposer for the Sequencer

Reviewer: ```suggestion
      Once transformed, the block data is validated to make sure that the proposer for the Sequencer
```
there isn't any commit/proposer set validation on the gossipnet block as it's the head-of-chain block which isn't committed/finalized yet

>       Once transformed, the block data is validated to make sure that the
>       validator set correctly agreed on the block, the proposer for the Sequencer
>       and the block match, and that the previous commit hash in the block header
>       is correct.

Reviewer: ```suggestion
      Once transformed, the block data is validated to make sure that the proposer for the block is the one expected. It also checks the commit of the parent block by verifying that >2/3 staking power of the sequencer chain voted for it.
```

there isn't any commit/proposer set validation on the gossipnet block as it's the head-of-chain block which isn't committed/finalized yet

      The block is then passed to the Executor actor as a
      `ExecutorCommand::BlockReceivedFromGossipNetwork` message which will
      ultimately process and filter the block.
- `DriverCommand::GetNewBlocks`([link](https://github.com/astriaorg/astria/blob/3c4e47dbe1818e4228691d6bfd2b2143a06f1a6e/crates/astria-conductor/src/driver.rs#L54))
    - This message triggers the sending of a `ReaderCommand::GetNewBlocks` to the
      Reader actor to initiate the pulling of data from the DA layer.

### Reader

> - Creates a `CelestiaClient` using the Celestia client implementation from `astria-sequencer-relayer` to communicate with the DA layer

Reviewer: ```suggestion
- Creates a `CelestiaClient` to communicate with the DA layer
```

gonna update this very soon (#27)

> - Creates a `TendermintClient` which is used when validating blocks from the DA

Reviewer: ```suggestion
- Creates a `TendermintClient` which is used when validating blocks 
```

> - Creates a `TendermintClient` which is used when validating blocks from the DA
>   layer against the Sequencer data

Reviewer: ```suggestion
- Creates a `TendermintClient` which is used when validating blocks received
```

- Runs an event loop that handles receiving `ReaderCommand`s that drive data retrieval from the DA layer
- Passes the blocks it receives to the `Executor`

The Reader receives a `ReaderCommand::GetNewBlocks`
([link](https://github.com/astriaorg/astria/blob/3c4e47dbe1818e4228691d6bfd2b2143a06f1a6e/crates/astria-conductor/src/driver.rs#L54))
message from the driver. The `CelestiaClient`
([link](https://github.com/astriaorg/astria/blob/3c4e47dbe1818e4228691d6bfd2b2143a06f1a6e/crates/astria-sequencer-relayer/src/data_availability.rs#L244))
is then called from the Reader to get data from the Celestia DA. This data is

> then parsed from a Celestia blob into individual blocks, each block is then validated to check that the validator set is correct and the

Reviewer: ```suggestion
then parsed from Celestia blobs into individual partial blocks (consisting of relevant information needed for validation + the relevant rollup transactions). Each block is then validated to check that the validator set is correct and the
```

> then parsed from a Celestia blob into individual blocks, each block is then validated to check that the validator set is correct and the
> proposer for the block matches the proposer of the Sequencer, and that the previous commit of the current block is correct.

Reviewer: ```suggestion
then parsed from a Celestia blob into individual blocks, Then, the block data is validated to make sure that the proposer for the block is the one expected. It also checks the commit of the parent block by verifying that >2/3 staking power of the sequencer chain voted for it.
```

Each block is then transformed into a `SequencerBlockSubset`s and handed off to the Executor as a

> `ExecutorCommand::BlockReceivedFromDataAvailability` for transaction filtering, and passing off to the rollup for execution.

Reviewer: the conductor has been updated to only pull relevant rollup txs, so there's no more transaction filtering

### Executor

> - Runs an event loop that handles receiving `ExecutorCommand`s from both the
>   Driver and Reader

Reviewer: ```suggestion
- Runs an event loop that handles receiving `ExecutorCommand`s from both the
  Driver and Reader.
```

- Filters transactions by their rollup namespace and sends them to the rollup for execution

> - Catalogs and matches the block hashes received from the gossip network and the DA layer by rollup namespace so that it can send `firm` commits to the rollup

Reviewer: ```suggestion
- Maps sequencer block hashes to execution block hashes so that it can send `firm` commits to the rollup
```

execution. The execution hash that is returned from the rollup is then stored in
a hash map for Sequencer block hash -> execution hash.

When blocks are received from the DA layer, the hash map of Sequencer block hash
-> execution hash is checked to see if the block has already passed through the
Conductor from the Sequencer. If the block isn't seen, it is filtered and sent
to the rollup for execution exactly the same way the transactions are sent when
received from the Sequencer, then a message to finalize the block is sent.
If the block is already present in the hash map, just the finalize block message
is sent. After being finalized, the Sequencer block hash -> execution hash entry
in the hash map is deleted.

## Execution Data

### Transaction Filtering

An instance of the Conductor is meant to be run alongside the rollup node.

The `chain_id` that Conductor uses as the rollup namespace is pulled from the rollup's config. As such, the Conductor is only aware of the single namespace of the rollup it is supporting. The `chain_id` or namespace is tracked in the Sequencer block that contains the transactions for the rollup and not in the transactions themselves.

> When a Sequencer block is received from the gossip network the Conductor filters the transactions for its namespace and passes only those transaction, and the previous execution hash from the rollup to the execution layer. See the [astria execution api](https://github.com/astriaorg/astria/blob/main/specs/execution-api.md) for more details.

Reviewer: ```suggestion
When a Sequencer block is received from the gossip network, the Conductor filters the transactions for its chain ID and executes only those transactions. See the [astria execution api](https://github.com/astriaorg/astria/blob/main/specs/execution-api.md) for more details.
```

`namespace` is celestia-specific and not part of sequencer blocks

### Data Validation

Data is validated before being sent to the rollup for execution. Validation occurs in two places:

- When blocks are received from the gossip network, the data is validated in Driver's `handle_network_event` using `BlockVerifier::validate_sequencer_block_data`

> - When blocks are fetched from the DA layer, the data is validated in Reader's `get_new_blocks` using `BlockVerifier::validate_signed_namespace_data` and `BlockVerifier::validate_rollup_data`

Reviewer: this changes after  #290 actually

Author: I'll create another branch to capture the changes for this!

Author: Created https://github.com/astriaorg/astria/pull/322

### Soft Commitments

> When a Sequencer block is received by the Conductor from the gossip network, it is assumed to be the `head` of the chain, thus its parent block is designated as a `soft` commitment. Data received in this manner has been validated by the Sequencer network, and rollups can trust that the data will not be reverted. This is also the primary moment in the data life cycle for Astria that transactions are sent to the rollup for execution.

Reviewer: ```suggestion
When a Sequencer block is received by the Conductor from the gossip network, it is assumed to be the `head` of the chain, thus its parent block is designated as a `soft` commitment. This is also the primary moment in the data life cycle for Astria that transactions are sent to the rollup for execution.
```

head blocks can still be reverted

As mentioned in the [Transaction Filtering](#transaction-filtering) section above, the only information sent to the rollup is the list of ordered transactions and the previous execution hash from the rollup. It is the rollup node's responsibility to build their own specific block from the data provided and return the execution hash that resulted from adding the new block.

The Conductor keeps a map of Sequencer block hashes to rollup execution hashes for later matching when blocks are seen in the DA layer.

### Firm Commitments

> When the Conductor pulls data from the DA, it again filters that data by namespace and compares the block hashes seen there with those of the already executed blocks stored in the map mentioned at the end of the [Soft Commitments](#soft-commitments) section. For each block seen in DA that matches an executed block, a `FinalizeBlock` message is sent to the rollup to set those blocks to `final` and the entries in the execution hash to Sequencer block hash map are cleared.

Reviewer: ```suggestion
When the Conductor pulls data from the DA, it compares the sequencer block hashes seen with those of the already executed blocks stored in the map mentioned at the end of the [Soft Commitments](#soft-commitments) section. For each block seen in DA that matches an executed block, a `FinalizeBlock` message is sent to the rollup to set those blocks to `final` and the entries in the execution hash to Sequencer block hash map are cleared.
```
